compiler/plugins/lazy_interp.ml

```diff
@@ -0,0 +1,273 @@
+(* This file is part of the Catala compiler, a specification language for tax
+   and social benefits computation rules. Copyright (C) 2023 Inria, contributor:
+   Louis Gesbert <louis.gesbert@inria.fr>.
+
+   Licensed under the Apache License, Version 2.0 (the "License"); you may not
+   use this file except in compliance with the License. You may obtain a copy of
+   the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+   WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+   License for the specific language governing permissions and limitations under
+   the License. *)
+
+open Catala_utils
+open Shared_ast
+
+(* -- Definition of the lazy interpreter -- *)
+
+let log fmt = Format.ifprintf Format.err_formatter (fmt ^^ "@\n")
+let error e = Errors.raise_spanned_error (Expr.pos e)
+let noassert = true
+
+type laziness_level = {
+  eval_struct : bool;
+      (* if true, evaluate members of structures, tuples, etc. *)
+  eval_op : bool;
+      (* if false, evaluate the operands but keep e.g. `3 + 4` as is *)
+  eval_default : bool;
+      (* if false, stop evaluating as soon as you can discriminate with
+         `EEmptyError` *)
+}
+
+let value_level = { eval_struct = false; eval_op = true; eval_default = true }
+
+module Env = struct
+  type 'm t =
+    | Env of
+        ((dcalc, 'm mark) gexpr, ((dcalc, 'm mark) gexpr * 'm t) ref) Var.Map.t
```

Comment:
I'm curious, could you put a comment describing what you're putting in the `Env.t` ?

Reply:
I will! Until then: an `env` maps a variable to an expression _and_ the corresponding captured env.
I don't want to think about how much more complicated that would have been in an imperative setting :D

The `ref` part is only used for laziness: the raw expression is first added to the environment, then we can replace it once computed. We don't use `lazy` because we want more control (in particular, we can reduce the same expression several times with different levels)